static/js/components/email/hoc.js

```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import R from 'ramda';
+
+import EmailCompositionDialog from './EmailCompositionDialog';
+import { getAllEmailProperties } from './lib';
+import { getDisplayName } from '../../util/util';
+
+export const withEmailDialog = R.curry((emailType, WrappedComponent) => {
+  class WithEmailDialog extends React.Component {
+    constructor(props) {
+      super(props);
+      this.renderCompositionDialog = this.renderCompositionDialog.bind(this);
+      this.allEmailProperties = getAllEmailProperties();
+      this.emailType = emailType;
+    }
+
+    renderCompositionDialog() {
+      let {
+        dispatch,
+        ui: { dialogVisibility },
+        email
+      } = this.props;
+      return <EmailCompositionDialog
+        dispatch={dispatch}
+        dialogVisibility={dialogVisibility}
+        email={email}
+        emailProperties={this.allEmailProperties}
+      />;
+    }
+
+    render() {
+      let emailDialog = this.props.emailDialog || this.renderCompositionDialog();
+
+      let emailOpenActions = this.props.emailOpenActions || {};
+      emailOpenActions[this.emailType] = R.partial(
+        this.allEmailProperties[this.emailType].emailOpenAction, [this.props.dispatch]
+      );
+
+      return <WrappedComponent
+        {...this.props}
+        emailDialog={emailDialog}
+        emailOpenActions={emailOpenActions}
+      />;
+    }
+  }
+
+  if (WithEmailDialog.displayName && !WithEmailDialog.displayName.startsWith('WithEmailDialogs')) {
+    WithEmailDialog.displayName = `WithEmailDialogs(${getDisplayName(WrappedComponent)})`;
+  }
+  return WithEmailDialog;
+});
```

Comment:
My understanding of the idea behind using an HOC for this was that we'd have something like a `WithEmail` HOC that handled dispatching all of the necessary actions, dealing with the API, etc. so that could be abstracted away from the rendering / UI code. 

It seems a bit of indirection to me to have the `EmailCompositionDialog`, which actually shoulders much of this responsibility, separate from this HOC. Does that make sense? I'm also not super into passing down `dispatch` as a prop - I think I've made my case for avoiding that pattern before though 😆